frame/staking/src/lib.rs

```diff
@@ -1384,7 +1782,185 @@ impl<T: Trait> Module<T> {
+		// that we need. Then it should be Self::validator_count(). Else it should be all the
+		// candidates.
+		let snapshot_length = <SnapshotValidators<T>>::decode_len()
+			.map_err(|_| Error::<T>::SnapshotUnavailable)?;
+		let desired_winners = Self::validator_count().min(snapshot_length as u32);
+		ensure!(winners.len() as u32 == desired_winners, Error::<T>::PhragmenBogusWinnerCount);
+
+		// decode snapshot validators.
+		let snapshot_validators = Self::snapshot_validators()
+			.ok_or(Error::<T>::SnapshotUnavailable)?;
+
+		// check if all winners were legit; this is rather cheap. Replace with accountId.
+		let winners = winners.into_iter().map(|widx| {
+			snapshot_validators.get(widx as usize).cloned().ok_or(Error::<T>::PhragmenBogusWinner)
+		}).collect::<Result<Vec<T::AccountId>, Error<T>>>()?;
+
+		// decode the rest of the snapshot.
+		let snapshot_nominators = <Module<T>>::snapshot_nominators()
+			.ok_or(Error::<T>::SnapshotUnavailable)?;
+
+		// helpers
+		let nominator_at = |i: NominatorIndex| -> Option<T::AccountId> {
+			// NOTE: we assume both `ValidatorIndex` and `NominatorIndex` are smaller than usize.
+			snapshot_nominators.get(i as usize).cloned()
+		};
+		let validator_at = |i: ValidatorIndex| -> Option<T::AccountId> {
+			// NOTE: we assume both `ValidatorIndex` and `NominatorIndex` are smaller than usize.
+			snapshot_validators.get(i as usize).cloned()
+		};
+
+		// un-compact.
+		let assignments = compact_assignments.into_assignment(
+			nominator_at,
+			validator_at,
+		).map_err(|e| {
+			debug::native::warn!(
+				target: "staking",
+				"un-compacting solution failed due to {:?}",
+				e,
+			);
+			Error::<T>::PhragmenBogusCompact
+		})?;
+
+		// check all nominators actually including the claimed vote. Also check correct self votes.
+		// Note that we assume all validators and nominators in `assignments` are properly bonded,
+		// because they are coming from the snapshot via a given index.
+		for Assignment { who, distribution } in assignments.iter() {
+			let is_validator = <Validators<T>>::contains_key(&who);
+			let maybe_nomination = Self::nominators(&who);
+
+			if !(maybe_nomination.is_some() ^ is_validator) {
+				// all of the indices must map to either a validator or a nominator. If this is ever
+				// not the case, then the locking system of staking is most likely faulty, or we
+				// have bigger problems.
+				debug::native::error!(
+					target: "staking",
+					"detected an error in the staking locking and snapshot."
+				);
+				// abort.
+				return Err(Error::<T>::PhragmenBogusNominator);
+			}
+
+			if !is_validator {
+				// a normal vote
+				let nomination = maybe_nomination.expect(
+					"exactly one of maybe_validator and maybe_nomination is true. \
+					is_validator is false; maybe_nomination is some; qed"
+				);
+				// NOTE: we don't really have to check here if the sum of all edges are the
+				// nominator correct. Un-compacting assures this by definition.
+				ensure!(
```

Comment:
for me here we are checking that the nominations inside the solution correspond a to a real nomination.
i.e. the nominator actually nominate the validator.

Reply:
We don't the comment is about checking the edge weights which is actually not checked. Instead here we just check edge nodes, that all voters and targets are correct. Since these two are hand in hand, i commented why we don't do the other one.